lib/sycamore/sycamore/connectors/qdrant/qdrant_writer.py

```diff
@@ -0,0 +1,132 @@
+from dataclasses import dataclass, asdict
+import typing
+from typing import Any, Optional, Union
+from typing_extensions import TypeGuard
+
+from sycamore.data.document import Document
+from sycamore.connectors.base_writer import BaseDBWriter
+from sycamore.utils.import_utils import requires_modules
+
+if typing.TYPE_CHECKING:
+    from qdrant_client import QdrantClient
+
+
+@dataclass
+class QdrantWriterClientParams(BaseDBWriter.ClientParams):
+    location: Optional[str] = None
+    url: Optional[str] = None
+    port: Optional[int] = 6333
+    grpc_port: int = 6334
+    prefer_grpc: bool = False
+    https: Optional[bool] = None
+    api_key: Optional[str] = None
+    prefix: Optional[str] = None
+    timeout: Optional[int] = None
+    host: Optional[str] = None
+    path: Optional[str] = None
+
+
+@dataclass
+class QdrantWriterTargetParams(BaseDBWriter.TargetParams):
+    collection_params: dict
+
+    def compatible_with(self, other: BaseDBWriter.TargetParams) -> bool:
+        assert isinstance(other, QdrantWriterTargetParams)
+
+        return all(
+            [
+                self.collection_params["collection_name"] == other.collection_params["collection_name"],
+                self.collection_params["vectors_config"] == other.collection_params["vectors_config"],
+            ]
+        )
+
+
+class QdrantWriterClient(BaseDBWriter.Client):
+    @requires_modules("qdrant_client", extra="qdrant")
+    def __init__(self, client: "QdrantClient"):
+        self._client = client
+
+    @classmethod
+    def from_client_params(cls, params: BaseDBWriter.ClientParams) -> "QdrantWriterClient":
+        assert isinstance(params, QdrantWriterClientParams)
+
+        from qdrant_client import QdrantClient
+
+        client = QdrantClient(**asdict(params))
+        return cls(client)
+
+    def write_many_records(self, records: list["BaseDBWriter.Record"], target_params: "BaseDBWriter.TargetParams"):
+        assert isinstance(target_params, QdrantWriterTargetParams)
+        assert _narrow_list_of_qdrant_records(records), f"Found bad records in {records}"
+
+        from qdrant_client.models import PointStruct
+
+        points = [PointStruct(id=record.id, vector=record.vector, payload=record.payload) for record in records]  # type: ignore
+
+        self._client.upload_points(collection_name=target_params.collection_params["collection_name"], points=points)
+
+    def create_target_idempotent(self, target_params: "BaseDBWriter.TargetParams"):
+        assert isinstance(target_params, QdrantWriterTargetParams)
+        params = target_params.collection_params
+        if not self._client.collection_exists(params["collection_name"]):
+            self._client.create_collection(**params)
```

Comment:
This is a race condition - if `create_collection` throws an error trying to create a collection that already exists then you need to handle it (swallow is fine - and you can actually swallow all errors here since we validate with get_existing_target_params + compatible_with)